tracer/src/Datadog.Trace/Ci/Agent/MessagePack/IEventMessagePackFormatter.cs

```diff
@@ -20,18 +20,25 @@ public IEvent Deserialize(byte[] bytes, int offset, IFormatterResolver formatter
 
         public int Serialize(ref byte[] bytes, int offset, IEvent value, IFormatterResolver formatterResolver)
         {
-            if (value is SpanEvent spanEvent)
-            {
-                return formatterResolver.GetFormatter<SpanEvent>().Serialize(ref bytes, offset, spanEvent, formatterResolver);
-            }
+            var formatter = CIFormatterResolver.Instance;
 
-            if (value is TestEvent testEvent)
+            if (value is CIVisibilityEvent<Span> ciVisibilityEvent)
             {
-                return formatterResolver.GetFormatter<TestEvent>().Serialize(ref bytes, offset, testEvent, formatterResolver);
+                if (formatter == formatterResolver)
+                {
+                    return formatter.GetCiVisibilityEventFormatter().Serialize(ref bytes, offset, ciVisibilityEvent, formatterResolver);
+                }
+
+                return formatterResolver.GetFormatter<CIVisibilityEvent<Span>>().Serialize(ref bytes, offset, ciVisibilityEvent, formatterResolver);
```

Comment:
Is there a reason that there are two code paths? Is the `formatter == formatterResolver` code path more efficient somehow?

Reply:
yes, is only to avoid all comparisons inside the `CIFormatterResolver` if I already knows the one we need to serialize the `value`. TBH I'm not saving so much here in terms of CPU cycles.